### PR TITLE
Logging: Add customizable backends

### PR DESCRIPTION
**Pull request by daniellimws**
This is part 2 of #3449 , also a follow-up of #3533

The main goal of this is to change the logging backend to support multiple sinks (console, color console, file, etc) through the Backend interface, with the following changes:

* Separate logging into its own thread
* Add support for adding and removing backends on runtime
* Change logging filter on runtime
* Button to open log location

As a side effect, on Windows, Citra has to be compiled as a GUI application to make the console hidden by default.

The current way of dealing with large log files is to set a limit of 50MB for logging to file. This should work fine when the log level is not Debug.

On Discord, we discussed about truncating repeated logs. For the aim of reducing the number of lines (spam on console, and also the number of bytes written to file) while still showing there is something being logged, we can print dots and when the message is finally changed show the number of times it was repeated. 

One example of a message being repeated 5 times would give something like (of course a better format is needed)

```
[  51.578907] Audio <Debug> audio_core/time_stretch.cpp:Process:58: Dropping frames!
.... (4)
```

Currently this has not been implemented yet, still listening to ideas as this will hide potentially important information like time.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3568)

### CODE REVIEW

**Comment by jroweboy** [comment]
https://github.com/citra-emu/citra/pull/3449/files#diff-4eaa9529af469fd20b11df19d2739172R146 <- gcc expects the `-mwindows` linker flag not the subsystem flag you are passing in there

**Comment by daniellimws** [comment]
Oh weird I think that change got lost in one of the rebases.

File: src/common/logging/backend.cpp [hunk 1]
```diff
@@ -4,17 +4,141 @@
...
+        std::lock_guard<std::mutex> lock(message_mutex);
+        message_queue.Push(std::move(e));
+        message_cv.notify_one();
+    }
+
+    void AddBackend(std::unique_ptr<Backend> backend) {
+        backends.push_back(std::move(backend));
+    }
+
+    void RemoveBackend(const std::string& backend_name) {
```
Comment (jroweboy) [inline src/common/logging/backend.cpp]: We should prevent issues when removing a backend thats currently writing by locking a mutex in the write_logs and acquiring this lock before removing or adding a backend.

Comment (daniellimws) [reply src/common/logging/backend.cpp]: Yup alright I'll add that.

**Comment by daniellimws** [comment]
Are we currently looking into logging to multiple files? If so the name of the FileBackend has to be dependent on the file name, to avoid removing all FileBackends at once with the current implementation.

**Comment by jroweboy** [comment]
@daniellimws I personally think its added complexity that we don't even know if we'll need yet. In my opinion, if we later decide that we need it we can add it, but theres no reason to require it for this part.

**Comment by daniellimws** [comment]
Yea that was what I thought. That's good then.

**Comment by jroweboy** [comment]
Will be merging this on the 7th if no one else reviews or merges it by then. It was already thoroughly reviewed when I PRed it, so it should still be ready to go 👍

**Comment by FearlessTobi** [comment]
@daniellimws Will you implement the log truncation for spammed lines in a follow-up PR?

**Comment by daniellimws** [comment]
Would like to listen to more opinions first, as there may be some things to consider, like
* What if the same log is printed a long time later, do we still truncate? Would the time information be important?
* What if the same 2 messages are printed continuously, do we care?

File: src/citra_qt/debugger/console.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include "citra_qt/debugger/console.h"
+#include "citra_qt/ui_settings.h"
+#include "common/logging/backend.h"
+
+namespace Debugger {
+void ToggleConsole() {
+#ifdef _WIN32
+    if (UISettings::values.show_console) {
+        if (AllocConsole()) {
+            freopen_s((FILE**)stdin, "CONIN$", "r", stdin);
```
Comment (wwylele) [inline src/citra_qt/debugger/console.cpp]: Are you sure this shouldn't be `freopen_s(&stdin, "CONIN$", "r", stdin);`?
and if yours is correct, please write it as 
`freopen_s(reinterpret_cast<FILE**>(stdin), "CONIN$", "r", stdin);`
as this is a strange pointer type casting.

File: src/common/file_util.h [hunk 3]
```diff
@@ -156,7 +156,7 @@ void SplitFilename83(const std::string& filename, std::array<char, 9>& short_nam
 class IOFile : public NonCopyable {
 public:
     IOFile();
-    IOFile(const std::string& filename, const char openmode[]);
+    IOFile(const std::string& filename, const char openmode[], int flags = 0);
```
Comment (wwylele) [inline src/common/file_util.h]: Please document what `flag` is used for.

File: src/common/logging/backend.cpp [hunk 4]
```diff
@@ -124,23 +251,34 @@ Entry CreateEntry(Class log_class, Level log_level, const char* filename, unsign
     entry.timestamp = duration_cast<std::chrono::microseconds>(steady_clock::now() - time_origin);
     entry.log_class = log_class;
     entry.log_level = log_level;
-    entry.filename = Common::TrimSourcePath(filename);
+    entry.filename = std::string(Common::TrimSourcePath(filename));
```
Comment (wwylele) [inline src/common/logging/backend.cpp]: Is this change intentional?

File: src/common/logging/backend.cpp [hunk 5]
```diff
@@ -124,23 +251,34 @@ Entry CreateEntry(Class log_class, Level log_level, const char* filename, unsign
...
     entry.line_num = line_nr;
-    entry.function = function;
-    entry.message = std::move(message);
+    entry.function = std::string(function);
+    entry.message = message;
```
Comment (wwylele) [inline src/common/logging/backend.cpp]: Is these two change intentional?

File: src/common/logging/backend.h [hunk 6]
```diff
@@ -6,8 +6,12 @@
 
 #include <chrono>
 #include <cstdarg>
+#include <memory>
 #include <string>
 #include <utility>
+#include <boost/optional.hpp>
```
Comment (wwylele) [inline src/common/logging/backend.h]: I don't see where this is needed.

**Comment by wwylele** [comment]
About log file truncation, I'd say it is fine as-is for this PR (50MB limit). We'll see how well it works and change the method in the future if needed.

File: src/common/file_util.h (hunk 3 above)
Comment (jroweboy) [reply src/common/file_util.h]: @daniellimws to help you document this, `flags` is used for windows specific file open mode flags. This allows the citra to open the logs in shared write mode, so that the file isn't considered "locked" while citra is open and people can open the log file and view it

File: src/citra_qt/debugger/console.cpp (hunk 2 above)
Comment (daniellimws) [reply src/citra_qt/debugger/console.cpp]: @jroweboy What about this?

File: src/common/logging/backend.h (hunk 6 above)
Comment (daniellimws) [reply src/common/logging/backend.h]: This was probably a mistake.

File: src/common/logging/backend.cpp (hunk 4 above)
Comment (daniellimws) [reply src/common/logging/backend.cpp]: Not sure why this is there, removing.

File: src/common/logging/backend.cpp (hunk 5 above)
Comment (daniellimws) [reply src/common/logging/backend.cpp]: Not sure why this is there, removing.

**Comment by daniellimws** [comment]
Oh crap there's some rebasing problem. I'll see what I can do.

**Comment by jroweboy** [comment]
@daniellimws @wwylele addressed the last small nitpicks, and pushed a rebased commit. please merge when you are ready. To answer your question about the stdin stuff, I did some research and updated the code there as well https://github.com/citra-emu/citra/pull/3568/files#diff-b636b0577a7bb569496810ed9d65ce58R21

**Comment by daniellimws** [comment]
Hmm I may have made a mistake...

**Comment by jroweboy** [comment]
@daniellimws i force pushed over your remote branch on github (sorry! was trying to help out with rebasing). if you wanted to keep working on it locally, you'd need to fetch the latest from the remote, and then hard reset to the remote. but now that you've pushed more commits, you can't do that. (you can just force push back to the old version if you needed)

**Comment by daniellimws** [comment]
@jroweboy No no I made a mistake :pensive:  could you do a forced push again really sorry

**Comment by jroweboy** [comment]
@wwylele i'm going to merge with the promise that i'll fix issues with it if they pop up
